### PR TITLE
[MM-60441] index channels after a user is added to a team

### PR DESCRIPTION
**Pull request by isacikgoz**
#### Summary
Right after creating the membership, we are going to index public channels for that team member.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-60441

#### Release Note

```release-note
Fix an issue where channels not showing up in the channel switcher after a user gets added to a team
```

### CODE REVIEW

File: server/channels/store/searchlayer/layer.go [hunk 1]
```diff
@@ -111,6 +111,28 @@ func (s *SearchStore) indexUser(rctx request.CTX, user *model.User) {
...
+					if err := engineCopy.IndexChannel(rctx, channel, []string{teamMember.UserId}, []string{teamMember.UserId}); err != nil {
+						rctx.Logger().Warn("Encountered error indexing channel", mlog.String("channel_id", channel.Id), mlog.String("search_engine", engineCopy.GetName()), mlog.Err(err))
+						return
+					}
+					rctx.Logger().Debug("Indexed channel in search engine", mlog.String("search_engine", engineCopy.GetName()), mlog.String("channel_id", channel.Id))
+				}
+			})
+		}
+	}
+}
```
Comment (agnivade) [inline server/channels/store/searchlayer/layer.go:134]: This does not look right. We need to actually get all team members for each channel, and then update the channel. We need to call `(c *SearchChannelStore) indexChannel` instead.

```go
for _, channel := range channels {
	s.channel.indexChannel(rctx, channel)
}
```

File: server/channels/store/searchlayer/layer.go [hunk 2]
```diff
@@ -111,6 +111,28 @@ func (s *SearchStore) indexUser(rctx request.CTX, user *model.User) {
 	}
 }
 
+func (s *SearchStore) indexChannelsForTeam(rctx request.CTX, teamMember *model.TeamMember) {
+	channels, err := s.Channel().GetPublicChannelsForTeam(teamMember.TeamId, 0, 100)
```
Comment (agnivade) [inline server/channels/store/searchlayer/layer.go:115]: We need to paginate and get all channels for the team. Otherwise this will work for only 100 channels.

File: server/channels/store/searchlayer/team_layer.go [hunk 3]
```diff
@@ -18,6 +18,7 @@ func (s SearchTeamStore) SaveMember(rctx request.CTX, teamMember *model.TeamMemb
 	member, err := s.TeamStore.SaveMember(rctx, teamMember, maxUsersPerTeam)
 	if err == nil {
 		s.rootStore.indexUserFromID(rctx, member.UserId)
+		s.rootStore.indexChannelsForTeam(rctx, member)
```
Comment (agnivade) [inline server/channels/store/searchlayer/team_layer.go:21]: We also need this for `RemoveMember` and `RemoveAllMembersByUser` as well right?

File: server/channels/store/searchlayer/layer.go [hunk 4]
```diff
@@ -111,6 +111,28 @@ func (s *SearchStore) indexUser(rctx request.CTX, user *model.User) {
...
+	if err != nil {
+		rctx.Logger().Warn("Encountered error while indexing public channels", mlog.String("team_member_id", teamMember.UserId), mlog.Err(err))
```
Comment (agnivade) [inline server/channels/store/searchlayer/layer.go:117]: Also string should be `Encountered error while getting all public channels for team` and we should log the team_id rather than the member_id. I don't believe the member_id has any role to play in the SQL query.

**Review by agnivade — CHANGES_REQUESTED** [review]
Left some comments.

File: server/channels/store/searchlayer/layer.go (hunk 1 above)
Comment (isacikgoz) [reply server/channels/store/searchlayer/layer.go:134]: TIL, so we always in theory purge other users when we re-index?

File: server/channels/store/searchlayer/team_layer.go (hunk 3 above)
Comment (isacikgoz) [reply server/channels/store/searchlayer/team_layer.go:21]: After your comment above, definitely :)

**Review by cpoile — APPROVED** [review]
I don't know the context around this, but should we be concerned about performance?

File: server/channels/store/searchlayer/layer.go (hunk 1 above)
Comment (agnivade) [reply server/channels/store/searchlayer/layer.go]: It's only channels. Yes, there is no other way because of the JSON structure and ES being a document database. The channel object contains the list of all team_members, so if a new member gets added to a team, all channel objects need to be updated.

File: server/channels/store/searchlayer/layer.go [hunk 5]
```diff
@@ -111,6 +111,32 @@ func (s *SearchStore) indexUser(rctx request.CTX, user *model.User) {
 	}
 }
 
+func (s *SearchStore) indexChannelsForTeam(rctx request.CTX, teamID string) {
```
Comment (harshilsharma63) [inline server/channels/store/searchlayer/layer.go:114]: Do we not index DMs and GMs or are they handled elsewhere?

File: server/channels/store/searchlayer/layer.go [hunk 6]
```diff
@@ -111,6 +111,32 @@ func (s *SearchStore) indexUser(rctx request.CTX, user *model.User) {
...
+	var (
+		page     int
+		perPage  = 100
+		channels []*model.Channel
+	)
+
+	for {
```
Comment (harshilsharma63) [inline server/channels/store/searchlayer/layer.go:121]: Unrelated to this PR but seems unnecessary to index all public channels of a team when a user joins or leaves a team. Instead we could index only the channel they joined by default (town Square, Off Topic and other channels defined in the config), and the same for only the channels a user was member of when they left.

Just curious as to the reasoning for this.

**Review by harshilsharma63 — APPROVED** [review]

File: server/channels/store/searchlayer/layer.go (hunk 5 above)
Comment (agnivade) [reply server/channels/store/searchlayer/layer.go:114]: DM/GMs are not tied to a team, so they don't need any change. Same for private channels. A user cannot access any private channels immediately after joining a team unless they are added to it.

**Comment by agnivade** [comment]
@isacikgoz - I am actually not able to repro the issue. My steps:
1. Do a fresh bulk index.
2. Add a user to a team.
3. Search for a public channel in the team for which the user is not a member of.

In master, I can see the search results.

If I query ES directly, I can actually see team_member_ids populated for the user:

```
curl -d '{"query":{"bool":{"filter":[{"term":{"team_member_ids":{"value":"uoao4pyo3jrpuxgdwn7f15btdc"}}},{"bool":{"minimum_should_match":1,"must":[{"prefix":{"name_suggestions":{"value":"mychan"}}}],"should":[{"bool":{"must_not":[{"term":{"type":{"value":"P"}}}]}},{"bool":{"filter":[{"term":{"user_ids":{"value":"uoao4pyo3jrpuxgdwn7f15btdc"}}}],"must":[{"term":{"type":{"value":"P"}}}]}}]}}]}},"size":50}' -H "Content-Type: application/vnd.elasticsearch+json" -s 'http://localhost:9200/channels/_search?typed_keys=true' | jq . | grep ywxebfjhspgd7y5zqab3q1f81r
            "ywxebfjhspgd7y5zqab3q1f81r"
```

What am I doing wrong?

**Comment by isacikgoz** [comment]
> I am actually not able to repro the issue.

Are you creating that user fresh? My reproduction steps were first creating the user, then add them to a team.

**Comment by agnivade** [comment]
Aha interesting! Let me do that. Thanks.

**Comment by agnivade** [comment]
Btw - I have some review comments for the PR. I just wanted to repro and understand the problem better before posting my review. And in the meanwhile, totally slammed with the major customer escalation today. Please allow me some time to get back on this. :)

**Review by agnivade — COMMENTED** [review]
I've finally managed to reproduce it. And your PR fixes the problem. Some suggestions to improve:

1. The `GetPublicChannelsForTeam` is a costly one if there are thousands of channels in that team. We should gate that operation only when a search backend is enabled:

```diff
diff --git a/server/channels/store/searchlayer/team_layer.go b/server/channels/store/searchlayer/team_layer.go
index 3738e3edb9..a6ddee28ae 100644
--- a/server/channels/store/searchlayer/team_layer.go
+++ b/server/channels/store/searchlayer/team_layer.go
@@ -17,8 +17,11 @@ type SearchTeamStore struct {
 func (s SearchTeamStore) SaveMember(rctx request.CTX, teamMember *model.TeamMember, maxUsersPerTeam int) (*model.TeamMember, error) {
        member, err := s.TeamStore.SaveMember(rctx, teamMember, maxUsersPerTeam)
        if err == nil {
-               s.rootStore.indexUserFromID(rctx, member.UserId)
-               s.rootStore.indexChannelsForTeam(rctx, member.TeamId)
+               // Nothing to do if search engine is not active
+               if s.rootStore.searchEngine.ActiveEngine() != "database" && s.rootStore.searchEngine.ActiveEngine() != "none" {
+                       s.rootStore.indexUserFromID(rctx, member.UserId)
+                       s.rootStore.indexChannelsForTeam(rctx, member.TeamId)
+               }
        }
        return member, err
 }
```

Please apply this to `RemoveMember` and `RemoveAllMembersByUser` as well.

2. Now I realize that this happens ONLY for a new user, not an existing user. This is what I don't have the answer to. Why does it make a difference if both users are not part of a team. How is the team_member_ids field populated for an existing user? Any idea on this?

3. This will likely have some impact in load tests (especially around the scenario where we join a huge number of users suddenly to a single team). Can we get some rough numbers on the cost of  `GetPublicChannelsForTeam` with page size of 100,200,500 and Channels table of size of 100K channels?

cc @agarciamontoro @streamer45  on point 3.

**Comment by agarciamontoro** [comment]
Let me use this as a testbed for a real PR :eyes:

**Comment by agarciamontoro** [comment]
The URL built by the pipeline is not correct! Sorry for the noise. WIll repeat when I know it works :)

**Comment by isacikgoz** [comment]
>     1. The `GetPublicChannelsForTeam` is a costly one if there are thousands of channels in that team. We should gate that operation only when a search backend is enabled:

Applied the suggest change, makes sense to me.

>     2. Now I realize that this happens ONLY for a new user, not an existing user. This is what I don't have the answer to. Why does it make a difference if both users are not part of a team. How is the team_member_ids field populated for an existing user? Any idea on this?

Indeed it is interesting, I'm thinking of there is a gap in the bulk index logic that somehow those channels getting indexed for he users that is not yet a member of the team. Maybe something to look separately?

**Comment by agnivade** [comment]
> Indeed it is interesting, I'm thinking of there is a gap in the bulk index logic that somehow those channels getting indexed for he users that is not yet a member of the team. Maybe something to look separately?

It slightly bothers me a bit that we haven't understood the issue completely. But agree that there is no need to block this PR. Wondering if we can open a HW ticket with all the details and let the community investigate it?

**Review by agnivade — APPROVED** [review]
Good to go, pending performance numbers for different page sizes of 100/200/500 and select something that is appropriate

**Comment by agnivade** [comment]
@isacikgoz - Can we run some checks for the ideal page size? We don't need a full-blown load-test here. I am just looking for the right value of the page_size.

File: server/channels/store/searchlayer/layer.go [hunk 7]
```diff
@@ -111,6 +111,32 @@ func (s *SearchStore) indexUser(rctx request.CTX, user *model.User) {
...
+		if err != nil {
+			rctx.Logger().Warn("Encountered error while retreiving public channels for indexing", mlog.String("team_id", teamID), mlog.Err(err))
+			return
+		}
+
+		channels = append(channels, cs...)
+
+		if len(channels) < perPage {
+			break
+		}
```
Comment (streamer45) [inline server/channels/store/searchlayer/layer.go:132]: This will need reviewing as we should check for `len(cs)` or the loop may never finish if `perPage` elements are returned on the first iteration. See https://hub.mattermost.com/xyz/pl/z5k46zi1m3g67d4adp4xgzye5o

One thought (to avoid missing this in the future) could be to build a small abstraction since this pattern is used in a few places and is likely to be re-used in the future. Example:

```go
func pager[T any](fetch func(page int) ([]T, error), perPage int) ([]T, error) {
	var list []T
	var page int

	for {
		fetched, err := fetch(page)
		if err != nil {
			return nil, err
		}

		list = append(list, fetched...)

		if len(fetched) < perPage {
			break
		}

		page++
	}

	return list, nil
}
```

An additional thought is that we may consider improving our dataset for unit tests since it's fairly common to have many channels, so clearly not an edge case.

Let me know what you think.

**Review by streamer45 — CHANGES_REQUESTED** [review]

**Comment by agarciamontoro** [comment]
Sorry for all the noise here, @isacikgoz, I feel we're really close to the end of this :nerd_face:

**Comment by isacikgoz** [comment]
> Sorry for all the noise here, @isacikgoz, I feel we're really close to the end of this 🤓

I'll add more commits to make your job harder :p

File: server/channels/store/searchlayer/layer.go (hunk 7 above)
Comment (isacikgoz) [reply server/channels/store/searchlayer/layer.go]: I love the idea of this function, created something in the `uitls` package, please take a look @streamer45

**Comment by agarciamontoro** [comment]
Juuuuust one more, I swear <3

File: server/public/utils/page_test.go [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+					return nil, errors.New("fetch error")
+				}
+				return []int{page*10 + 1, page*10 + 2, page*10 + 3}, nil
+			},
+			perPage:   3,
+			expected:  []int{1, 2, 3},
+			expectErr: true,
+		},
+		{
+			name: "fetch with fewer items than perPage",
```
Comment (streamer45) [inline server/public/utils/page_test.go:45]: Can we add one case for a page size larger than the available elements?

File: server/channels/store/searchlayer/team_layer.go [hunk 9]
```diff
@@ -33,15 +37,30 @@ func (s SearchTeamStore) UpdateMember(rctx request.CTX, teamMember *model.TeamMe
...
 	return err
 }
 
 func (s SearchTeamStore) RemoveAllMembersByUser(rctx request.CTX, userId string) error {
-	err := s.TeamStore.RemoveAllMembersByUser(rctx, userId)
+	memberships, err := s.TeamStore.GetTeamsForUser(rctx, userId, "", true)
+	if err != nil {
+		return err
+	}
+	if s.rootStore.searchEngine.ActiveEngine() != "database" && s.rootStore.searchEngine.ActiveEngine() != "none" {
```
Comment (streamer45) [inline server/channels/store/searchlayer/team_layer.go:54]: Don't we have these strings ("database", "none") encoded in some constants or type?

File: server/channels/store/searchlayer/layer.go [hunk 10]
```diff
@@ -111,6 +112,25 @@ func (s *SearchStore) indexUser(rctx request.CTX, user *model.User) {
...
+	channels, err := utils.Pager(func(page int) ([]*model.Channel, error) {
+		return s.channel.GetPublicChannelsForTeam(teamID, page, perPage)
+	}, perPage)
+	if err != nil {
+		rctx.Logger().Warn("Encountered error while retreiving public channels for indexing", mlog.String("team_id", teamID), mlog.Err(err))
+		return
+	}
+
+	for _, channel := range channels {
+		s.channel.indexChannel(rctx, channel)
```
Comment (streamer45) [inline server/channels/store/searchlayer/layer.go:130]: Curious to understand why this `indexChannel` method doesn't return an error.

**Review by streamer45 — APPROVED** [review]
Thanks @isacikgoz , looks great!

**Comment by agnivade** [comment]
> -2.50217959895379250218%

Too few decimal places. We need more precision!.

**Review by cpoile — APPROVED** [review]

**Comment by agarciamontoro** [comment]
Ah, new timeouts to fix, how fun! :sob:

**Comment by agarciamontoro** [comment]
I swear this is the last one.

**Comment by agarciamontoro** [comment]
Taking a look at this, I want to run some tests.

**Comment by agarciamontoro** [comment]
Mini-report: this is what a single user triggers when joining a new team, in a server with the 100M-posts DB, with no users connected:
image
image


The problem is not the call to `GetPublicChannelsForTeam` (although I think that may become worse at scale), but the call to `GetTeamMembersForChannel` in `indexChannel`.

image

When a user joins a team T, the server retrieves all the team members in team T *for every public channel in team T*. That's a lot of data in a large database. Improving this should be easy: pre-compute all team members in T, and then pass that data to every `indexChannel`. Will that alone fix this? Let's see.

**Comment by agarciamontoro** [comment]
It just killed a `c7i.8xlarge` :joy: I'll go back to testing with smaller DBs, I guess.

**Comment by agnivade** [comment]
See my original note here: https://github.com/mattermost/mattermost/pull/29165#pullrequestreview-2435503898. (point 2). There's something in the code which we don't understand yet. (atleast I don't). There shouldn't be a difference with new user added to team, vs existing user. I think if we can understand this, we should be able to figure it out.

> It just killed a c7i.8xlarge 😂 I'll go back to testing with smaller DBs, I guess.

You can just set a `LIMIT 500` in the SQL query or whatever. Up to you.

**Comment by agarciamontoro** [comment]
This seems to be a replica lag issue! I've confirmed with @agnivade that he tested this with just one DB node, and I've been testing with two this whole time. I just switched to using only 1 node and the issue went away. Now searching for `15hh` returns the channel :)

This is explained by these two sequential events:

1. From `SearchTeamStore.SaveMember`, we first call `SqlTeamStore.SaveMember`, which ends up making the `INSERT INTO TeamMembers ...` query, adding the new member.
2. After that, if Elasticsearch is enabled, we call `indexChannelsForTeam`, which calls `indexChannel`, which calls `GetTeamMembersForChannel`, which ends up making the `SELECT tm.UserId FROM Channels c, Teams t, TeamMembers tm ...` query, retrieving the members of the team.

If the replica has not caught up between events 1 and 2, then it will retrieve an outdated list of team members, thus updating the channel in the index with that old list.

**Comment by agarciamontoro** [comment]
Closing in favour of #33400.
